src/validator/IndexScanValidator.cpp

```diff
@@ -93,26 +95,173 @@ Status IndexScanValidator::prepareFilter() {
     }
 
     auto *filter = sentence->whereClause()->filter();
-    auto ret = checkFilter(filter, *sentence->from());
-    NG_RETURN_IF_ERROR(ret);
     storage::cpp2::IndexQueryContext ctx;
-    ctx.set_filter(Expression::encode(*filter));
+    if (needTextSearch(filter)) {
+        NG_RETURN_IF_ERROR(checkTSService());
+        if (!textSearchReady_) {
+            return Status::Error("Text search service not ready");
+        }
+        auto retExpr = rewriteTSFilter(filter);
+        if (!retExpr.ok()) {
+            return retExpr.status();
+        }
+        if (isEmptyResultSet_) {
+            // return empty result direct.
+            return Status::OK();
+        }
+        auto ret = checkFilter(retExpr.value().get());
+        NG_RETURN_IF_ERROR(ret);
+        ctx.set_filter(Expression::encode(*std::move(retExpr.value()).get()));
+    } else {
+        auto ret = checkFilter(filter);
+        NG_RETURN_IF_ERROR(ret);
+        ctx.set_filter(Expression::encode(*filter));
+    }
     contexts_ = std::make_unique<std::vector<storage::cpp2::IndexQueryContext>>();
     contexts_->emplace_back(std::move(ctx));
     return Status::OK();
 }
 
-Status IndexScanValidator::checkFilter(Expression* expr, const std::string& from) {
+StatusOr<std::unique_ptr<Expression>>
+IndexScanValidator::rewriteTSFilter(Expression* expr) {
+    std::vector<std::string> values;
+    auto tsExpr = static_cast<TextSearchExpression*>(expr);
+    auto vRet = textSearch(tsExpr);
+    if (!vRet.ok()) {
+        return Status::Error("Text search error.");
+    }
+    if (vRet.value().empty()) {
+        isEmptyResultSet_ = true;
+        return Status::OK();
+    } else if (vRet.value().size() == 1) {
+        auto relExpr = std::make_unique<RelationalExpression>(
+            Expression::Kind::kRelEQ,
+            new LabelAttributeExpression(new LabelExpression(*tsExpr->arg()->from()),
+                                         new LabelExpression(*tsExpr->arg()->prop())),
+            new ConstantExpression(Value(vRet.value()[0])));
+        return relExpr;
+    } else {
+        return genOrFilterFromList(tsExpr, vRet.value());
+    }
+    return Status::OK();
+}
+
+std::unique_ptr<Expression> IndexScanValidator::genOrFilterFromList(
+    TextSearchExpression* expr, const std::vector<std::string>& values) {
+    auto filter = std::make_unique<LogicalExpression>(
```

Comment:
It seems `LogicalExpression` has support multiple operands, could we use it?

Reply:
BTW, have test the performance between pass as filter or pass by  multi`IndexQueryContext`